### PR TITLE
dev-cmd/extract: instruct users to run git fetch unshallow

### LINKED ISSUE #7315: `brew extract` should instruct users to run `git fetch --unshallow`
# A detailed description of the proposed feature

When `brew extract` fails to find a previous version of a formula, it fails with:
https://github.com/Homebrew/brew/blob/e12da9491d2dc557636458c265196fedde34a30a/Library/Homebrew/dev-cmd/extract.rb#L135

However, it can fail to find the desired version simply because `hombrew-core` is shallow, and running `git -C $(brew --repository homebrew/homebrew-core) fetch --unshallow` before retrying `brew extract` simply solves the problem.

# The motivation for the feature

For people that are not aware that Homebrew uses Git, or what even is a shallow clone, it could be great to have the error message instruct them to try to "unshallow" the tap repository.

# How the feature would be relevant to at least 90% of Homebrew users

I don't know about 90%, but I'd say this would be helpful for 90% of `brew extract` users!

# What alternatives to the feature have been considered

`brew extract` could run `git fetch --unshallow` itself, but I guess that some people would not like that because of disk space.

### PR DESCRIPTION
**Pull request by vidusheeamoli**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Resolves #7315.

### CODE REVIEW

File: Library/Homebrew/dev-cmd/extract.rb [hunk 1]
```diff
@@ -132,7 +132,15 @@ def extract
...
-        odie "Could not find #{name}! The formula or version may not have existed." if rev.nil?
+        if rev.nil? && (source_tap.path/".git/shallow").exist?
+          opoo <<~EOS
+            #{source_tap} is shallow clone. Run
+              git -C $(brew --repository homebrew/homebrew-core) fetch --unshallow
+              and retry.
+          EOS
+        elsif rev.nil?
+          odie "Could not find #{name}! The formula or version may not have existed."
+        end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/extract.rb:143]:
```suggestion
        if rev.nil? && source_tap.shallow?
          opoo <<~EOS
            #{source_tap} is shallow clone! Retry after running:
              git -C "$(#{source_tap.path})" fetch --unshallow
          EOS
        elsif rev.nil?
          odie "Could not find #{name}! The formula or version may not have existed."
        end
```

Comment (vidusheeamoli) [reply Library/Homebrew/dev-cmd/extract.rb:143]: On it! @MikeMcQuaid  😄

Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/extract.rb:143]: @vidusheeamoli Thanks! Note you can commit straight from the GitHub UI if desired 😎

File: Library/Homebrew/dev-cmd/extract.rb [hunk 2]
```diff
@@ -132,7 +132,14 @@ def extract
...
         rev, (path,) = Git.last_revision_commit_of_files(repo, pattern, before_commit: rev)
-        odie "Could not find #{name}! The formula or version may not have existed." if rev.nil?
+        if rev.nil? && source_tap.shallow?
+          opoo <<~EOS
+            #{source_tap} is shallow clone! Retry after running:
+              git -C "$(#{source_tap.path})" fetch --unshallow
+          EOS
+        elsif rev.nil?
+          odie "Could not find #{name}! The formula or version may not have existed."
+        end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/extract.rb:142]:
```suggestion
        if rev.nil? && source_tap.shallow?
          odie <<~EOS
            Could not find #{name} but #{source_tap} is a shallow clone! 
            Try again after running:
              git -C "$(#{source_tap.path})" fetch --unshallow
          EOS
        elsif rev.nil?
          odie "Could not find #{name}! The formula or version may not have existed."
        end
```

Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/extract.rb:142]: Whoops, noticed another tweak: this should still also `odie` and say it wasn't found!

File: Library/Homebrew/dev-cmd/extract.rb [hunk 3]
```diff
@@ -132,7 +132,15 @@ def extract
       loop do
         rev = rev.nil? ? "HEAD" : "#{rev}~1"
         rev, (path,) = Git.last_revision_commit_of_files(repo, pattern, before_commit: rev)
-        odie "Could not find #{name}! The formula or version may not have existed." if rev.nil?
+        if rev.nil? && source_tap.shallow?
+          odie <<~EOS
+            Could not find #{name} but #{source_tap} is a shallow clone! 
+            Try again after running:
+              git -C "$(#{source_tap.path})" fetch --unshallow
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/extract.rb:139]:
```suggestion
              git -C "#{source_tap.path}" fetch --unshallow
```

Sorry, I'm an idiot! 😭. Last change, I promise!

Comment (vidusheeamoli) [reply Library/Homebrew/dev-cmd/extract.rb:139]: Noo! Thank you for helping! 😃

**Comment by MikeMcQuaid** [comment]
Thanks again for another great contribution @vidusheeamoli!

**Comment by phil-blain** [comment]
thanks @vidusheeamoli ! Since I don't know Ruby at all it's great you were able to quickly implement this fix!

**Pull request merged**